src/unity/python/turicreate/toolkits/_mps_utils.py

```diff
@@ -418,57 +471,85 @@ def init(self, n, c_in, h_in, w_in, c_out, h_out, w_out, config=None, weights=No
-    # input layer.
-    def wait_for_batch(self):
-        if self._mode == MpsGraphMode.Train:
-            self._LIB.TCMPSWaitForTrainingBatchGraph(self.handle, _ctypes.byref(self._buf_loss))
-            loss = _np.frombuffer(self._buf_loss, dtype=_np.float32).copy()
-            return loss
-        elif self._mode == MpsGraphMode.TrainReturnGrad:
-            self._LIB.TCMPSWaitForTrainReturnGradBatchGraph(self.handle, _ctypes.byref(self._buf_out_fp16))
-            raw_out = _np.frombuffer(self._buf_out_fp16, dtype=_np.float16)
-            out = raw_out.reshape(self._ishape).astype(_np.float32).copy()
-            return out
-        else:
-            self._LIB.TCMPSWaitForInferenceBatchGraph(self.handle, _ctypes.byref(self._buf_out_fp16))
-            raw_out = _np.frombuffer(self._buf_out_fp16, dtype=_np.float16)
-            out = raw_out.reshape(self._oshape).astype(_np.float32).copy()
-            return out
+    def predict(self, input):
+        """
+        Submits an input batch to the model. Returns a MpsFloatArray
+        representing the model predictions. Calling asnumpy() on this value will
+        wait for the batch to finish and yield the predictions as a numpy array.
+        """
+
+        assert self._mode == MpsGraphMode.Inference
+        assert input.shape == self._ishape
+
+        input_array = MpsFloatArray(input)
+        result_handle = _ctypes.c_void_p()
+        status_code = self._LIB.TCMPSPredictGraph(
+            self.handle, input_array.handle, _ctypes.byref(result_handle))
+
+        assert status_code == 0, "Error calling TCMPSPredictGraph"
+        assert result_handle, "TCMPSPredictGraph unexpectedly returned NULL pointer"
+
+        result = MpsFloatArray(result_handle)
+        assert result.shape() == self._oshape
+
+        return result
+
+    def train_return_grad(self, input, grad):
+        """
+        Performs a forward pass from the input batch, followed by a backward
+        pass using the provided gradient (in place of a loss function). Returns
+        a MpsFloatArray representing the output (final gradient) of the backward
+        pass. Calling asnumpy() on this value will wait for the batch to finish
+        and yield the output as a numpy array.
+        """
+
+        assert self._mode == MpsGraphMode.TrainReturnGrad
+        assert input.shape == self._ishape
+        assert grad.shape == self._oshape
+
+        input_array = MpsFloatArray(input)
+        grad_array = MpsFloatArray(grad)
+        result_handle = _ctypes.c_void_p()
+        status_code = self._LIB.TCMPSTrainGraph(
+            self.handle, input_array.handle, grad_array.handle,
+                _ctypes.byref(result_handle))
+
+        assert status_code == 0, "Error calling TCMPSTrainReturnGradGraph"
+        assert result_handle, "TCMPSTrainReturnGradGraph unexpectedly returned NULL pointer"
+
+        result = MpsFloatArray(result_handle)
+        assert result.shape() == self._ishape
+
+        return result
```

Comment:
A few questions about `train_return_grad`. Are we using this for anything other than testing? Passing out gradients is only necessary for hybrid models (part MPS, part something else). I'm just curious if this is actually used or planned to be used anywhere. Of course, testing is very important too.

Can I also get the loss here? It seems to me that in an ideal world, there would only be `train(..., return_input_gradient=False)`, where you could flip on `return_input_gradient` to have it return a tuple of `(loss, grad)`, instead of just `loss`. I'm calling it "input" gradient here, so as not to imply it returns all gradients.

Reply:
Yeah, good question. Honestly, I only included `train_return_grad` because I've been erring on the side of preserving functionality until I can take a serious pass on adding tests here (using these new interfaces). Although ideally I would love to be able to write tests that just use the "normal" interfaces for training/inference. I also want to give more thought to what we need for debugging future networks; honestly my focus right now is on adapting OD.